Online payments made SIMPLE - How to work with Stripe (#110)
rootstrap/blog: content/How to work with Stripe.md

Cons

 - Expensive (high % fee)

## 2) INSTALLATION

This post assumes you already created a Stripe account and so you have access to dashboard and its configuration.

#### RAILS

- Add these two gems:
    - [Stripe](https://github.com/stripe/stripe-ruby) to achieve the integration
    - [Stripe Testing](https://github.com/stripe-ruby-mock/stripe-ruby-mock) to test your integration, you don't want to end up writing lots of mocking classes, right?
- Configure your keys & version from the Stripe dashboard

https://gist.github.com/fedeagripa/0ae76f1f2a5b77f0a12fdbc2eeaa5904

#### REACT

- Add this package [Stripe](https://github.com/stripe/react-stripe-elements)
- Configure your App to use the same api key as for rails (make sure it is the same, as you start moving between envs you may forget it). Remember that there is a testing key and a live one.

Add an env file to store your keys
https://gist.github.com/fedeagripa/5f99c824308ba899d6f7deb5fb85ff76

Add your Stripe wrapper

https://gist.github.com/fedeagripa/f2f64556a820abf9b377fa16f256d0d2

## 3) START USING PAYMENTS WITH STRIPE

### CREDIT CARDS

##### REACT - DO YOURSELF A FAVOR AND USE THE EXISTING COMPONENT

> I'm not a fan of reinventing the wheel by any means, the design this components provide are more than enough for 99% of the apps you will be building. But if you insist, be prepared to spend 2 weeks dealing with details instead of 2 days.

Reviewer: ```suggestion
I'm not a fan of reinventing the wheel by any means, the design these components provide is more than enough for 99% of the apps you will be building. But if you insist, be prepared to spend 2 weeks dealing with details instead of 2 days.
```

https://gist.github.com/fedeagripa/385cb4c4f9ccf66f833749349a41426a

##### RAILS - DON'T TRY TO STORE ALL THE INFO (IT'S ILEGAL)

You will tend to store more credit card info that you need. The only info that you need to store in your database (for basic usage) is:
  - `customer_id` : Stripe customer identifier that you will store in your User for example
  - `card_id` : Stripe card identifier

The `token_id` you will get from your frontend is a short lived token that is only needed for an atomic operation.

Add a `customer_id` field to your user (or Shop in next example).
Add a `card_id` to your user (or Shop in next example).

Now take this service example (Shopify page example):
https://gist.github.com/fedeagripa/478e2c3a1e5eb018341740504fcb272f

And this simple controller:

https://gist.github.com/fedeagripa/29f02be2c8ab39560d06de47e13c7681

And that's all! You can start charging your users now!
All fraud detections and customer service actions can be managed directly from Stripe's dashboard.

### SUBSCRIPTIONS

> To create a subscription you need to define it, then create a product in Stripe (this last one is really clear looking at the dashboard, so i'm not going to explain it)

Reviewer: ```suggestion
To create a subscription you need to define it, then create a product in Stripe (this last one is really clear looking at the dashboard, so I'm not going to explain it)
```

##### CREATING THE SUBSCRIPTION
https://gist.github.com/fedeagripa/538df087957a2019e8444417347a32ac

> In that model you will store attribures like: `expires_at`, `type` or even `provider` if later you want to extend to other providres like PayPal or Apple Pay

Reviewer 2: typo in `providers`

Finally to create them on Stripe is quite simple:

https://gist.github.com/fedeagripa/d49a024299d2b8e315ff0861d32e9534

### COUPONS
Coupons are the abstract concept of `30% off` for example, when you apply that coupon to a user that's called a `discount`.
So you should define some discounts on Stripe and store their ids in your database to apply them to users.

> There are two types of coupons `percentage` & `fixed amount`, and any of them can be one time only or have the capability to be applied multiple times. So when you try to apply a coupon to a subscription for example, remember that it can fail if you reached the maximum usage number.

Reviewer: ```suggestion
There are two types of coupons `percentage` & `fixed amount`, and any of them can be one time only or have the capability to be applied multiple times. So when you try to apply a coupon to a subscription, for example, remember that it can fail if you reached the maximum usage number.
```

Another useful case that is worth mentioning is to apply a coupon to a user, this means that they will have a positive balance for any future invoice (be careful if you charge users with multiple products)

### SUBSCRIPTION ITEMS

> These are your billing items, so for the case of a web subscription you will just have 1 subscription item. For specific cases like an amazon cart or any complicated use case (where you have multiple items being added to a purchase) is where you have to start considering adding some specific logic to your app.

Reviewer: ```suggestion
These are your billing items, so for the case of a web subscription, you will just have 1 subscription item. For specific cases like an amazon cart or any complicated use case (where you have multiple items being added to purchase) is where you have to start considering adding some specific logic to your app.
```

I won't get really into detail about this, I just wanted to show the general concept behind this, maybe I will write more in detail in a future post.

##### RENEWALS
Don't overthink it, there is a webhook for most of your use cases. But for this specific need you can configure the following events:

 - customer.subscription.updated
This event happens every time a susbscription is updated according to this [documentation](https://stripe.com/docs/billing/subscriptions/change)

 - customer.subscription.deleted

> As simple as it sounds, it tells you when a subscription is canceled so you can take the actions needed in your app (possibly dissable the associated account)

Reviewer: ```suggestion
As simple as it sounds, it tells you when a subscription is canceled so you can take the actions needed in your app (possibly disable the associated account)
```

 - invoice.payment_succeeded

> This is a really important one! It tells us when a payment is actually accepted by the credit card provider (some times they can be fraud or declined)

Reviewer: ```suggestion
This is a really important one! It tells us when payment is actually accepted by the credit card provider (some times there can be fraud or the payment could get declined)
```

### WEBHOOKS
There are a lot of them and they will solve most of your problems, the only downcase is the headache trying to understand which exactly to use.

> I'm sorry to dissapoint you if you reached here trying to answer this question but up to now I only know [this page](https://stripe.com/docs/api/events/types) that explains the different existing webhooks and what they do. The other option is when you go to create a webhook from the developer's Stripe dashboard, they explain a bit more in detail what each event does.

Reviewer: ```suggestion
I'm sorry to disappoint you if you reached here trying to answer this question but up to now I only know [this page](https://stripe.com/docs/api/events/types) that explains the different existing webhooks and what they do. The other option is when you go to create a webhook from the developer's Stripe dashboard, they explain a bit more in detail what each event does.
```

> I'm sorry to dissapoint you if you reached here trying to answer this question but up to now I only know [this page](https://stripe.com/docs/api/events/types) that explains the different existing webhooks and what they do. The other option is when you go to create a webhook from the developer's Stripe dashboard, they explain a bit more in detail what each event does.

Reviewer: ```suggestion
I'm sorry to disappoint you if you reached here trying to answer this question but up to now I only know [this page](https://stripe.com/docs/api/events/types) that explains the different existing webhooks and what they do. The other option is when you go to create a webhook from the developer's Stripe dashboard, they explain a bit more in detail what each event does.
```

## 4) SPECIAL RECOMMENDATIONS
Keep these Stripe documentation pages as your friends:
 - [Devs api](https://stripe.com/docs/api/)

>  - [Events types](https://stripe.com/docs/api/events/types)

Reviewer 2: You're sharing the same page here and above (line 120). Just in case this wasn't intended

Author: It was, as stripe doc is a mess and when you cmd + F they launch a custom search I decided to shared different sections besides being the same page

Sometimes there are two or even three ways of solving a problem, so consider this and take your time to analyze each requirement properly before you start coding.

## 5) CONCLUSIONS

> You can easily add online payments to your app and test it in just 1 week (or so), thats amazing! The other amazing thing is that you can start managing most of the daily based situations like fraud of disputes just from the dashboard (you don't need to keep coding).

Reviewer: ```suggestion
You can easily add online payments to your app and test it in just 1 week (or so), that's amazing! The other amazing thing is that you can start managing most of the daily based situations like fraud of disputes just from the dashboard (you don't need to keep coding).
```

> The difficult part of this is when you start adding more concrete and detailed transactions and supporting multiple transfers types (like bank account tranfers instead of just visa or master). So if you liked this post and want to know more don't hesitate to leave some comments asking for it!

Reviewer 2: Maybe `... just Visa or MasterCard`?

> The difficult part of this is when you start adding more concrete and detailed transactions and supporting multiple transfers types (like bank account tranfers instead of just Visa or MasterCard). So if you liked this post and want to know more don't hesitate to leave some comments asking for it!

Reviewer: ```suggestion
The difficult part of this is when you start adding more concrete and detailed transactions and supporting multiple transfer types (like bank account transfers instead of just Visa or MasterCard). So if you liked this post and want to know more don't hesitate to leave some comments asking for it!
```

> # **Online payments done SIMPLE - How to work with Stripe**

Reviewer 3: I think I would use `made` here, instead of `done`. You make payments, instead of do payments.

![main image](images/payment.jpg)

Here at [Rootstrap](https://www.rootstrap.com/mobile-app-development-los-angeles/), we care about not re inventing the wheel and making things simple (along with high quality of course).

In this blog post, you’ll learn how start working with stripe to add in a fast way online payments to your apps.

## 1) WHY STRIPE?

Pros

 - Easy to implement and use

>  - Fast to implement, so your client will be happy

Reviewer 3: Maybe `Fast to develop`? So you can emphasize that it's fast and avoid `implement`

 - Solves most of your usual payment problems, so you don't lose time or clients (even worst)

 - Amazing dashboard with lot of capabilities so your client financial team can work along with you

Cons

 - Expensive (high % fee)

## 2) INSTALLATION

This post assumes you already created a stripe account and so you have access to dashboard and it configuration.

#### RAILS

> - Add this two gems

Reviewer 3: And `these`, not `this`, plural

    - [Stripe](https://github.com/stripe/stripe-ruby) to achieve the integration

>     - [Stripe Testing](https://github.com/stripe-ruby-mock/stripe-ruby-mock) to test your integration, you don't want to end up writting lots of mocking classes, right?

Reviewer 3: `Writing`, without the double `t`

- Configure your keys & version from the stripe dashboard

``` ruby
# config/intializers/stripe.rb

Rails.configuration.stripe = {
  publishable_key: ENV['STRIPE_PUBLISHABLE_KEY'],
  secret_key: ENV['STRIPE_SECRET_KEY']
}

Stripe.api_key = Rails.configuration.stripe[:secret_key]

```

#### REACT

- Add this package [Stripe](https://github.com/stripe/react-stripe-elements)
- Configure your App to use the same api key as for rails (make sure it is the same, as you start moving between envs you may forget it)

Add your Stripe wrapper

> ``` js

Reviewer 3: I see there are a lot of code examples here, do you think you could put them on gists and reference them here instead of having the code in this file?

Reviewer 3: It's pretty straightforward, you only have to create one under your user, copy and paste this code and put the link here.

class BillingForm extends Component {
  async onSubmit(result) {
    const { stripe, submitBilling, shopId, initialValues } = this.props;
    const data = result.toJS();

    /* AT THIS POINT THE CC IS CREATED AT STRIPE AND YOU NEED TO TELL YOUR BACKEND ABOUT IT */
    const { token, error } = await stripe.createToken(decamelizeKeys(data));

    if (error) {
      throw new SubmissionError({
        _error: error.message
      });
    }

    /* HERE WE WERE SUBMITING AND LENDING THE INFO TO THE BACKEND */
    await submitBilling(shopId, camelizeKeys(token), initialValues.get('name'));
  }

  render() {
    /* all your consts */
    return (
      ....
      <form onSubmit={handleSubmit(this.onSubmit)} className="p-3">
        /* The rest of your user profile form */
        /* CC real info */
        <div className="col-lg-3 offset-1">
          <div className="form-group">
            <label className="form-control-label">Card On File</label>
            <div>{brand && last4 ? `${brand} ending in ${last4}` : 'none'}</div>
          </div>
          <div className="form-group">
            <label className="form-control-label">Card Number</label>

>             <CardNumberElement key={`cardNumber${cardInputKey}`} className="form-control" />

Reviewer 3: I think some of the variables in the render should be on the part where he put `/* all your consts*/` maybe a better explanation can be added so it's more clear.

          </div>
          <div className="form-group">
            <label className="form-control-label">Expiry</label>
            <CardExpiryElement key={`cardExpiry${cardInputKey}`} className="form-control wd-80" />
          </div>
          <div className="form-group">
            <label className="form-control-label">CVC</label>
            <CardCVCElement key={`cardCvc${cardInputKey}`} className="form-control wd-80" />
          </div>
        </div>
      </form>
    )
  }
}

export default injectStripe(reduxForm({
  ...
})(BillingForm));
```

> ##### RAILS - DONT TRY TO STORE ALL INFO (ITS ILEGAL)

Reviewer 3: And `DON'T`

    }
    @customer = Stripe::Customer.create(customer_params)
    shop.update!(customer_id: @customer.id)
  end
end
```

And this simple controller:

```ruby
# app/controllers/api/v1/credit_cards_controller.rb
module Api
  module V1
    class CreditCardsController < Api::V1::ApiController
      helper_method :shop

      def index
        service = StripeService.new(shop)
        @card = service.credit_card_info
        @customer = service.customer
      end

      def create
        StripeService.new(shop).add_card(token_id, email, customer_name) if token_id
        head :no_content
      end

      def update
        begin
          StripeService.new(shop).update_credit_card(token_id)
        rescue StripeService::StripeException => ex
          return render_not_found(ex.message)
        end
        head :no_content
      end

      private

      def shop
        @shop ||= current_shop
      end

      def token_json
        params[:token]
      end

      def token_id
        token_json['id']
      end

      def email
        token_json['email']
      end

      def customer_name
        token_json['name']
      end
    end
  end
end
```

> And thats all! you can start charging your users now!

Reviewer 3: And after the exclamation mark you finished a sentence, so I would capitalize the you (`You`)

All fraud detections and customer service actions can be managed directly from stripe dashboard.

### SUBSCRIPTIONS
To create a subscription you need to define them in your app and a product in stripe (this last one is really clear looking at dashboard, so i'm not going to explain it)

> ##### CREATING THEM

Reviewer 3: Maybe be more explicit here, instead of `THEM` say `CREATING THE/A SUBSCRIPTION`. Titles are easily seen if a reader takes a quick look at the article and they wouldn't understand what `THEM` means if they don't read it all.

```ruby
# app/models/subscription.rb
class Subscription < ActiveRecord::Base
  belongs_to :user
  belongs_to :purchase_plan # this can be optional if you have annual or monthly plans for example
  has_many :subscription_items, dependent: :destroy # I'm going to explain this later

  enum status: ['define_your_possible_statuses']
end
```

In that model you will store attribures like: `expires_at`, `type` or even `provider` if later want to extend to other providres like paypal or apple pay

Finally to create them on stripe is quite simple

```ruby
# app/services/stripe_service.rb

def create_subscription
  Stripe::Subscription.create(
    customer: customer.id,
    plan: subs_plan_id, # this is the id of your plan (eg: monthly, annual, etc)

>     coupon: discount_code # if you have any (check discounts below section to understand it in more detail)

Reviewer 3: Hahahaha the coupon/discount problem comes from Masterclass 😛

>     coupon: discount_code # if you have any (check discounts below section to understand it in more detail)

Author: I still have nightmares with that naming problem 😢

  )
end
```

##### RENEWALS
Don't overthink it, there is a webhook for most of your cases. But for this specific need you can configure the following events

 - customer.subscription.updated

> This event happens every time a susbscription is update according to this [docuemntation](https://stripe.com/docs/billing/subscriptions/change)

Reviewer 3: There is a typo in `documentation` also

 - customer.subscription.deleted

> As simple as it sounds, it tells you when a subscription is canceled so you can take the actions needed in you app (possible disable account or so)

Reviewer 3: `your app`, missing the `r`

 - invoice.payment_succeeded
This is a really important one! it tell us when a payment is actually accepted by the credit card provider (some times they can be fraud or declined)

### SUBSCRIPTION ITEMS

> This are your bill items, so for the case of a web subscription you will just have 1 subscription item. For specific cases like a amazon cart or any complicated case you can imagine is where you have to start considering add some specific logic to your app.

Reviewer 3: Yeah, and I would try not to repeat `specific` in the same sentence.

> This are your bill items, so for the case of a web subscription you will just have 1 subscription item. For specific cases like a amazon cart or any complicated case you can imagine is where you have to start considering add some specific logic to your app.

Author: I changed it a bit, my intention was to explain that a cart has multiple items generally

I won't get really into detail about this, I just wanted to show the general concept behind this, maybe I will write more in detail in a future post.

### COUPONS

> Coupons are the abstract concept of `30% off` for example, later when you apply that coupon to a user thats called `discount`.

Reviewer 3: And I would remove the `later`

So you should define some discounts at stripe and store their ids in your database to apply them to users.
There is two types of coupons `percentage` & `fixed amount`, and any of them can be one time only or have the capability to be applied multiple times. So when you try ton apply a coupon to a subscription for example, remember that it can fail if you reached the maximum usage number.

Another useful case that worth to be explained is to apply a coupon to a user, this means that they will have a positive balance for any future invoice (be careful if you charge users with multiple products)

### WEBHOOKS
There is a lot of them and they will solve most of your problems, the only downcase is the headache trying to understand which exactly to use.

> I'm sorry to dissapoint you if you reached here trying to answer this question, but up to now I only know [this page](https://stripe.com/docs/api/events/types), and when you go to create a webhook from developers stripe dashboard they explain a bit more in detail what each event does.

Reviewer 3: I wouldn't use `as of now`, it means `from now on` https://www.collinsdictionary.com/dictionary/english-thesaurus/as-of-now I think `up to now` or `up until now` are ok. But the sentence is a little confusing too. You only know that page for what, I think you have to be more explicit there, like `but up to now this is the only page I know that discusses this issue/problem` or something like that, but there's something missing from there.

## 4) SPECIAL RECOMMENDATIONS

> Keep this stripe documentation pages as your friends:

Reviewer 3: `These`, you're talking about the pages.

 - [Devs api](https://stripe.com/docs/api/)
 - [Events types](https://stripe.com/docs/api/events/types)

 Sometimes there is two or even three ways of solving a problem, so consider this and take your time to analyze each requirement properly before start coding.

## 5) CONCLUSIONS
You can easily add online payments to your app and test it in just 1 week (or so), thats amazing!. The other amazing thing is that you can start managing most of the daily based situations like fraud of disputes just from the dashboard (you don't need to keep coding).

> The difficult part of this is when you start adding more concrete and detailed transactions and supporting multiple payemnt types (like bank account tranfers instead of just visa or master). So if you liked this post and want to know more don't hesitate to leave some comments asking for it!

Reviewer 3: `transfers`
